core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -348,12 +352,16 @@ class Partition(val topic: String,
    * 1. Partition ISR changed
    * 2. Any replica's LEO changed
    *
+   * We only increase HW if HW is smaller than LEO of all replicas whose lag <= replicaLagTimeMaxMs.
+   * This means that if a replica does not lag much behind leader and the replica's LEO is smaller than HW, HW will
+   * wait for this replica to catch up so that this replica can be added to ISR set.
+   *
    * Returns true if the HW was incremented, and false otherwise.
    * Note There is no need to acquire the leaderIsrUpdate lock here
    * since all callers of this private API acquire that lock
    */
-  private def maybeIncrementLeaderHW(leaderReplica: Replica): Boolean = {
-    val allLogEndOffsets = inSyncReplicas.map(_.logEndOffset)
+  private def maybeIncrementLeaderHW(leaderReplica: Replica, curTime: Long = time.milliseconds): Boolean = {
+    val allLogEndOffsets = assignedReplicas.filter(curTime - _.lastCaughtUpTimeMs <= replicaManager.config.replicaLagTimeMaxMs).map(_.logEndOffset)
```

Comment:
Ah I see you point. You are right, it may be subset of ISR because maybeShrink() is called periodically. I will fix this problem by explicitly including ISR in consideration here.